### ISSUE TITLE
should `podman inspect` require configuration of CNI network plugin?

### ISSUE THREAD

**Issue by miabbott** [containers/libpod#2384]
/kind bug

TL;DR - doing `podman inspect` requires configuration of the CNI plugin; has dependencies on `nsenter`, `iptables`, and `ip6tables`.  Is this setup really necessary for the `inspect` operation?

---

For Red Hat CoreOS (RHCOS) we are trying to compose the OS, build QCOWs, and create `oscontainers` all within an unprivileged container.  To do some of this, we are using a `supermin` appliance to run some of the commands that require a full host.

When configuring the `supermin` appliance, we specify the RPMs from the host that we want included in the appliance.  I configured the appliance to pull in `podman`, `buildah`, and `skopeo`.  After building a container in the appliance using `buildah`, I tried to inspect the container with `podman inspect`.  This failed initially because it needed `nsenter` in the path.  But what was more surprising was that `podman` required to setup the CNI plugin to inspect a container on disk:

```
error creating libpod runtime: error configuring CNI network plugin: exec: "nsenter": executable file not found in $PATH
Traceback (most recent call last):
  File "/usr/lib/coreos-assembler/oscontainer.py", line 157, in <module>
    tls_verify=not args.disable_tls_verify)
  File "/usr/lib/coreos-assembler/oscontainer.py", line 117, in oscontainer_build
    inspect = run_get_json(['podman', rootarg, 'inspect', image_name_and_tag])[0]
  File "/usr/lib/coreos-assembler/oscontainer.py", line 19, in run_get_json
    return json.loads(subprocess.check_output(args))
  File "/usr/lib64/python3.7/subprocess.py", line 395, in check_output
    **kwargs).stdout
  File "/usr/lib64/python3.7/subprocess.py", line 487, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['podman', '--root=/srv/cache/container-work/containers-storage', 'inspect', 'myoscontainer']' returned non-zero exit status 125.
```

When I got my `PATH` sorted out, I hit further errors when `podman` bailed out because it couldn't run `iptables` and `ip6tables`.  (This was due to `/etc/alternatives/` not being populated correctly, FWIW)

Anyways, I wanted to get thoughts on the need to configure the CNI plugin for something like `podman inspect`.  Granted this use case in a very slimmed down environment is unique, but it seems like extra overhead for `podman` to require things like `iptables` to just inspect a container.

Thanks!

**Comment by mheon** [comment]
We don't strictly need CNI to inspect the container, but it's necessary for a lot of other operations (create, start, stop, etc). As such, we configure it as part of initializing Podman.

It would be theoretically possible to make something that could inspect containers without CNI (which is what really brings in iptables, ip6tables, nsenter), but at that point we aren't really talking about Podman. We've removed the ability to do... just about anything except `ps`, `inspect`, maybe `pause`/`unpause`. Some of the pod operations too, but you won't be able to make containers to go in the pods, so they don't have much point.

**Comment by miabbott** [comment]
> It would be theoretically possible to make something that could inspect containers without CNI (which is what really brings in iptables, ip6tables, nsenter), but at that point we aren't really talking about Podman.

Something like `skopeo` perhaps?  :grin: 

I tried to use that for this use case, but it doesn't like interacting with the non-default containers-storage.  See https://github.com/containers/skopeo/issues/547#issuecomment-465397396

**Comment by rhatdan** [comment]
Would buildah inspect give you what you needed?  Were you inspecting a container or an image?

**Comment by miabbott** [comment]
> Would buildah inspect give you what you needed? Were you inspecting a container or an image?

I was inspecting a container image.  

I looked at the output of `buildah inspect` and the format is much different from that `podman` spits out.  Whereas, the format of `skopeo inspect` is closer to that of `podman inspect`, but more brief.

Our usecase is looking to retrieve the `Digest` and `Labels` specifically.

**Comment by rhatdan** [comment]
Does skopeo inspect get those for you?

**Comment by miabbott** [comment]
> Does skopeo inspect get those for you?

It does, but as I stated in my [previous comment](https://github.com/containers/libpod/issues/2384#issuecomment-465700551), I'm unable to use it in this case because containers storage is somewhere other than `/var/lib/containers/storage`

(Sorry for the delayed response, I was missing email notifications from GH for a while)

**Comment by rhatdan** [comment]
We have just merged some enhancements to skopeo to use containers storage configuration files.  Does this help? @vrothberg WDYT?

**Comment by vrothberg** [comment]
Yes, the latest master of Skopeo supports inspecting images in different locations (including rootless inspection).

**Comment by baude** [comment]
im going to close this ... @miabbott re-open if you still want podman to handle this.
